contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -320,6 +348,583 @@ babelfish_concat_wrapper(PG_FUNCTION_ARGS)
+		{
+			return (int)df_tz;
+		}
+		else
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				errmsg("\'%s\' is not a recognized datepart option",field)));
+			
+			return -1;
+		}
+	}
+
+	PG_CATCH();
+
+	{
+		PG_RE_THROW();	
+	}
+
+	PG_END_TRY();
+	return 1;
+}
+
+
+/*
+ * datepart_internal_datetimeoffset takes datetimeoffset and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_datetimeoffset(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	Timestamp	timestamp;
+	int			df_tz = PG_GETARG_INT32(2);
+
+	timestamp = (Timestamp)(DirectFunctionCall1(common_utility_plugin_ptr->datetimeoffset_timestamp,
+							PG_GETARG_DATUM(1)));
+	
+	timestamp = timestamp + (int64) df_tz * SECS_PER_MINUTE * USECS_PER_SEC;
+	
+	PG_RETURN_INT32(datepart_internal(field, timestamp, (float8)df_tz, false));
+}
+
+/*
+ * datepart_internal_date takes date and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_date(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	Timestamp	timestamp;
+	int			df_tz = PG_GETARG_INT32(2);
+	DateADT		date_arg;
+	date_arg = PG_GETARG_DATEADT(1);
+	timestamp = DirectFunctionCall1(date_timestamptz, date_arg);
+
+	PG_RETURN_INT32(datepart_internal(field, timestamp, (float8)df_tz, false));
+}
+
+/*
+ * datepart_internal_datetime takes datetime and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_datetime(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	Timestamp	timestamp;
+	int			df_tz = PG_GETARG_INT32(2);
+	
+	timestamp = PG_GETARG_TIMESTAMPTZ(1);
+
+	PG_RETURN_INT32(datepart_internal(field, timestamp, (float8)df_tz, false));
+}
+
+/*
+ * datepart_internal_int takes int and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_int(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	int64		num = PG_GETARG_INT64(1);
+	int			result;
+
+	/* 
+	 * Setting the timestamp in datepart_internal as 0 and passing num in third argument 
+	 * as there is no need of df_tz
+	 */
+	result = datepart_internal(field, 0, num, true);
+
+	PG_RETURN_INT32(result);
+}
+
+/*
+ * datepart_internal_money takes money and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_money(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	int64		num = PG_GETARG_INT64(1);
+	int			result;
+
+	/* 
+	 * Setting the timestamp in datepart_internal as 0 and passing num in third argument 
+	 * as there is no need of df_tz
+	 */
+	result = datepart_internal(field, 0, (float8)num/10000, true);
+
+	PG_RETURN_INT32(result);
+}
+
+/*
+ * datepart_internal_double takes double and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_double(PG_FUNCTION_ARGS)
```

Comment:
This function is not being linked with any SQL. Why is this?
Do we need this function? We shall use float8

Also here you are not doing ceil -1
rather you are doing result value - 1 - seems redundant